p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,540 @@
+	if err != nil {
+		return nil, fmt.Errorf("generate key for cert: %w", err)
+	}
+	cert, err := webrtc.GenerateCertificate(pk)
+	if err != nil {
+		return nil, fmt.Errorf("generate certificate: %w", err)
+	}
+	config := webrtc.Configuration{
+		Certificates: []webrtc.Certificate{*cert},
+	}
+	noiseTpt, err := noise.New(noise.ID, privKey, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create noise transport: %w", err)
+	}
+	transport := &WebRTCTransport{
+		rcmgr:        rcmgr,
+		webrtcConfig: config,
+		privKey:      privKey,
+		noiseTpt:     noiseTpt,
+		localPeerId:  localPeerID,
+
+		peerConnectionTimeouts: ICETimeouts{
+			Disconnect: DefaultDisconnectedTimeout,
+			Failed:     DefaultFailedTimeout,
+			Keepalive:  DefaultKeepaliveTimeout,
+		},
+
+		maxInFlightConnections: DefaultMaxInFlightConnections,
+	}
+	for _, opt := range opts {
+		if err := opt(transport); err != nil {
+			return nil, err
+		}
+	}
+	return transport, nil
+}
+
+func (t *WebRTCTransport) Protocols() []int {
+	return []int{ma.P_WEBRTC}
+}
+
+func (t *WebRTCTransport) Proxy() bool {
+	return false
+}
+
+func (t *WebRTCTransport) CanDial(addr ma.Multiaddr) bool {
+	return dialMatcher.Matches(addr)
+}
+
+var webRTCMultiAddr = ma.StringCast("/webrtc")
+
+func (t *WebRTCTransport) Listen(addr ma.Multiaddr) (tpt.Listener, error) {
+	addr, wrtcComponent := ma.SplitLast(addr)
+	isWebrtc := wrtcComponent.Equal(webRTCMultiAddr)
+	if !isWebrtc {
+		return nil, fmt.Errorf("must listen on webrtc multiaddr")
+	}
+	nw, host, err := manet.DialArgs(addr)
+	if err != nil {
+		return nil, fmt.Errorf("listener could not fetch dialargs: %w", err)
+	}
+	udpAddr, err := net.ResolveUDPAddr(nw, host)
+	if err != nil {
+		return nil, fmt.Errorf("listener could not resolve udp address: %w", err)
+	}
+
+	socket, err := net.ListenUDP(nw, udpAddr)
+	if err != nil {
+		return nil, fmt.Errorf("listen on udp: %w", err)
+	}
+
+	listener, err := t.listenSocket(socket)
+	if err != nil {
+		socket.Close()
+		return nil, err
+	}
+	return listener, nil
+}
+
+func (t *WebRTCTransport) listenSocket(socket *net.UDPConn) (tpt.Listener, error) {
+	listenerMultiaddr, err := manet.FromNetAddr(socket.LocalAddr())
+	if err != nil {
+		return nil, err
+	}
+
+	listenerFingerprint, err := t.getCertificateFingerprint()
+	if err != nil {
+		return nil, err
+	}
+
+	encodedLocalFingerprint, err := internal.EncodeDTLSFingerprint(listenerFingerprint)
+	if err != nil {
+		return nil, err
+	}
+
+	certMultiaddress, err := ma.NewMultiaddr(fmt.Sprintf("/webrtc/certhash/%s", encodedLocalFingerprint))
+	if err != nil {
+		return nil, err
+	}
+
+	listenerMultiaddr = listenerMultiaddr.Encapsulate(certMultiaddress)
+
+	return newListener(
+		t,
+		listenerMultiaddr,
+		socket,
+		t.webrtcConfig,
+	)
+}
+
+func (t *WebRTCTransport) Dial(ctx context.Context, remoteMultiaddr ma.Multiaddr, p peer.ID) (tpt.CapableConn, error) {
+	scope, err := t.rcmgr.OpenConnection(network.DirOutbound, false, remoteMultiaddr)
```

Comment:
This is a bit subtle, but we're actually using a FD here. When I open two WebRTC from one peer to two different remote peers, these connections originate from different UDP ports, which means that Pion opened new UDP sockets.

Reply:
What do you mean with FD?

Reply:
The second parameter to open connection (what you passed `false` to) is whether this connection uses a fd (file descriptor)

Reply:
Ok thanks for that, modified it.